helix-core/src/movement.rs

```diff
@@ -1,184 +1,254 @@
+            let end = slice.line_to_char(line + 1).saturating_sub(1);
+            nth_next_grapheme_boundary(slice, pos, count).min(end)
         }
     };
-    Range::new(if extend { range.anchor } else { pos }, pos)
+    let anchor = match behaviour {
+        Movement::Extend => range.anchor,
+        Movement::Move => pos,
+    };
+    Range::new(anchor, pos)
 }
 
 pub fn move_vertically(
-    text: RopeSlice,
+    slice: RopeSlice,
     range: Range,
     dir: Direction,
     count: usize,
-    extend: bool,
+    behaviour: Movement,
 ) -> Range {
-    let Position { row, col } = coords_at_pos(text, range.head);
+    let Position { row, col } = coords_at_pos(slice, range.head);
 
     let horiz = range.horiz.unwrap_or(col as u32);
 
     let new_line = match dir {
         Direction::Backward => row.saturating_sub(count),
         Direction::Forward => std::cmp::min(
             row.saturating_add(count),
-            text.len_lines().saturating_sub(2),
+            slice.len_lines().saturating_sub(2),
         ),
     };
 
     // convert to 0-indexed, subtract another 1 because len_chars() counts \n
-    let new_line_len = text.line(new_line).len_chars().saturating_sub(2);
+    let new_line_len = slice.line(new_line).len_chars().saturating_sub(2);
 
     let new_col = std::cmp::min(horiz as usize, new_line_len);
 
-    let pos = pos_at_coords(text, Position::new(new_line, new_col));
+    let pos = pos_at_coords(slice, Position::new(new_line, new_col));
+
+    let anchor = match behaviour {
+        Movement::Extend => range.anchor,
+        Movement::Move => pos,
+    };
 
-    let mut range = Range::new(if extend { range.anchor } else { pos }, pos);
+    let mut range = Range::new(anchor, pos);
     range.horiz = Some(horiz);
     range
 }
 
-pub fn move_next_word_start(slice: RopeSlice, mut begin: usize, count: usize) -> Option<Range> {
-    let mut end = begin;
-
-    for _ in 0..count {
-        if begin + 1 == slice.len_chars() {
-            return None;
-        }
-
-        let mut ch = slice.char(begin);
-        let next = slice.char(begin + 1);
-
-        // if we're at the end of a word, or on whitespce right before new one
-        if categorize(ch) != categorize(next) {
-            begin += 1;
-        }
-
-        if !skip_over_next(slice, &mut begin, |ch| ch == '\n') {
-            return None;
-        };
-        ch = slice.char(begin);
-
-        end = begin + 1;
-
-        if is_word(ch) {
-            skip_over_next(slice, &mut end, is_word);
-        } else if is_punctuation(ch) {
-            skip_over_next(slice, &mut end, is_punctuation);
-        }
-
-        skip_over_next(slice, &mut end, char::is_whitespace);
-    }
-
-    Some(Range::new(begin, end - 1))
+pub fn move_next_word_start(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Option<Range> {
+        let after_head = (range.head + 1).skip_newlines(slice);
+        (after_head + 1).inside(slice).then(|| {
```

Comment:
I think `then` is a bit tool overkill, feel like we are doing code golf, it's harder to understand compared to `if cond { } else { }`. Also, what does the `?` at the end do?

Reply:
Personally, I find
```
if cond {
       Some(value)
   } else {
       None
   }
```

Wordier and less idiomatic than
```rust
condition.then(|| Some(value))
```
I don't feel there's any code golfing going on there, but I'm happy to subject it to consensus if other people weigh in :).

As for the `?`, it's needed to break out of the closure with a `None` when we've reached the end of the slice, and thus don't want to make any further anchor moves. This `?` is specifically what fixes the previous bug of moving the anchor when already having selected the last word.

Reply:
I have fairly limited Rust knowledge at this point given I've only started learning a couple of days ago, but I do plenty of experience working with monads in other languages such as Haskell, Elm and the dry-monads in Ruby.

Personally, I agree with @PabloMansanet here. `.then` is common and I also tend to prefer it over `if/else`.

Reply:
I think `.then` is perfectly fine, it was only added recently to Rust so it's not as widely used yet.

Reply:
> As for the ?, it's needed to break out of the closure with a None when we've reached the end of the slice, and thus don't want to make any further anchor moves. This ? is specifically what fixes the previous bug of moving the anchor when already having selected the last word.

Why not rather than using `?`, we `unwrap_or(range)` and just return `Range`?

Reply:
> > As for the ?, it's needed to break out of the closure with a None when we've reached the end of the slice, and thus don't want to make any further anchor moves. This ? is specifically what fixes the previous bug of moving the anchor when already having selected the last word.
> 
> Why not rather than using `?`, we `unwrap_or(range)` and just return `Range`?

I like that, it might be more readable that way. Gonna check that it doesn't break anything but it looks good :+1: